### ISSUE TITLE
namespaces for executables do not work

### ISSUE THREAD

**Issue by skull-squadron** [erikhuda/thor#249]
Simply ripping off two examples from the wiki and putting them together...
## Code

``` ruby
#!/usr/bin/env ruby
require "rubygems" # ruby1.9 doesn't "require" it though
require "thor"

class Blah < Thor
  namespace :bar
  desc :foo, "Prints foo"
  def foo
    puts "foo"
  end
end

Blah.start
```
## With Thor :smile:

``` shell
ᐅ thor list
Could not find task "list" in "bar" namespace.
bar

---
thor bar:foo  # Prints foo
```
## As Executable :fearful:

``` shell
ᐅ ./blah.thor 
Tasks:
  blah.thor foo          # Prints foo
  blah.thor help [TASK]  # Describe available tasks or one specific task
```
## As Executable (Expected)

``` shell
ᐅ ./blah.thor 
Tasks:
  blah.thor bar:foo      # Prints foo
  blah.thor help [TASK]  # Describe available tasks or one specific task
```

**Comment by mod** [comment]
Bump, me too

**Comment by kaelumania** [comment]
Same here. It seems that the namespace syntax is only working with the thor executable correctly, but not with a standalone script.

I get the following
```
$ be ./bin/tool
Commands:
  tool config ACTION   # Manipulate environment configuration
  tool help [COMMAND]  # Describe available commands or one specific command
```

with a subcommand like

```
be ./bin/tool help config
Commands:
  tool config get KEY         # get the environment configuration
  tool config set KEY VALUE   # set the environment configuration
```

But what i really want is the following:
```
be ./bin/tool help config
Commands:
  tool config:get KEY         # get the environment configuration
  tool config:set KEY VALUE   # set the environment configuration
```

Any ideas to achieve that? The wiki seems to concentrate on the task handling part of thor and not the CLI interface programming anymore.

**Comment by taylorzr** [comment]
Bump.

**Comment by NicosKaralis** [comment]
10 years, any update on this?

I really want to move our base from rake to thor but this is the only thing keeping us

**Comment by JoelTowell** [comment]
I believe this is the same underlying issue as [#792](https://github.com/rails/thor/issues/792). [This](https://github.com/rails/thor/issues/792#issuecomment-1671830149) comment illustrates the runner-style usage, the tasks should be defined in a `Thorfile`, should not call `Blah.start`, and should be invoked via `thor`. 

The hacks in [this](https://github.com/rails/thor/issues/249#issuecomment-9701995) comment essentially recreate the runner entry point. For one, setting `$thor_runner = true` means `namespace` is set to `true` in the `formatted_usage` [method](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor/command.rb#L42-L49) when it is called from the `banner` [method](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor.rb#L546-L550) (note that `$thor_runner` is passed as the value for `namespace`). This gets you the desired `help` output, but doesn't influence command lookup. In `Thor::Runner`, `method_missing` [calls](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor/runner.rb#L37) `Thor::Util.find_class_and_command_by_namespace`, which [splits on](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor/util.rb#L131-L147) `':'` to derive the command. This is why calling  `Thor::Runner.start` instead of `Blah.start` will dispatch the namespaced command correctly.

It is important to notice that `bin/thor` does the [exact same](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/bin/thor) things as those hacks. The upshot here is that we already get the desired behaviour when invoking tasks in a `Thorfile` using the `thor` command.

The issue is that you are trying to invoke your `Thor` class as an executable using `Blah.start`. In this case, `$thor_runner` is `false` due to [this line](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor.rb#L4). In fact, the [specs are explicit](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/spec/thor_spec.rb#L389-L393) that `help` omits namespaces when `$thor_runner` is `false`. Additionally, direct `start` [calls](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor/base.rb#L583-L585) the `dispatch` method, which [does not call](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/lib/thor.rb#L504-L528) `Thor::Util.find_class_and_command_by_namespace`, so the namespaced command is never resolved via this route. The specs for `Thor::Runner` show it [working with namespaces](https://github.com/rails/thor/blob/cb11e8e87b3cfe6bc9f6d4b092897f0dfda2476d/spec/runner_spec.rb#L62-L75).

TL;DR: this conflates two different uses of `Thor`, invoking as an executable with `ThorKlass.start`, and invoking commands in a `Thorfile` using the `thor` executable. I don't think this is a bug, and it can probably be closed.
